### DISCUSSION TITLE
How to force a view to re-render when data has changed remotely?

### DISCUSSION THREAD

**Discussion by aloukissas** [remix-run/remix#4533] [Q&A]
I may have missed something huge here but here goes.

I'm playing with the jokes app and when I have the app running in two different browsers, the one browser doesn't know of the changes made by the other (e.g. added a new joke or deleted one), unless I hard-refresh. In other server-side web frameworks I've worked with (e.g. Phoenix LiveView), this is handled by a websocket connection that pushes changes to other clients; other approaches would do an http2 server push. With client-side rendering, one would use something like `swr` with long-poll or firebase web SDK to keep the view in sync with remote data.

Is this even supported today out of the box with Remix? If not, I'd expect this to be in the FAQ, tbh.

[edit] I see there's an existing [discussion](https://github.com/remix-run/remix/discussions/2622) on this topic already.

**Comment by machour** [comment]
Ryan teased something yesterday 👀
https://mobile.twitter.com/ryanflorence/status/1588898617301270530

**Reply by aloukissas** [reply]
I was able to do something like this with the existing APIs:

```typescript
import { useState, useEffect } from "react";
import { useFetcher, useLoaderData } from "@remix-run/react";

const DEFAULT_REFRESH_INTERVAL = 5_000;

interface Props {
  refreshInterval?: number;
}

export const useRouteData = <T,>({ refreshInterval }: Props) => {
  const loaderData = useLoaderData<T>();
  const fetcher = useFetcher();
  const [data, setData] = useState(loaderData);

  // Update data state from the loader on load, after form submission, etc.
  useEffect(() => {
    setData(loaderData);
  }, [loaderData]);

  // Get new data every refreshInterval msec (or every 5 sec by default)
  useEffect(() => {
    if (typeof window === "undefined") {
      return;
    }

    const interval = setInterval(() => {
      if (window.document.visibilityState === "visible") {
        fetcher.load(window.location.pathname);
      }
    }, refreshInterval || DEFAULT_REFRESH_INTERVAL);

    return () => clearInterval(interval);
  }, [fetcher, refreshInterval]);

  // When fetcher has new data, update the data state
  useEffect(() => {
    if (fetcher.data) {
      setData(fetcher.data);
    }
  }, [fetcher.data]);

  return { data, fetcher };
};
```
